cpp/src/io/orc/writer_impl.cu

```diff
@@ -1189,13 +1189,40 @@ orc_table_view make_orc_table_view(table_view const& table,
   }
 
   rmm::device_uvector<orc_column_device_view> d_orc_columns(orc_columns.size(), stream);
+  using stack_value_type = thrust::pair<column_device_view const*, thrust::optional<uint32_t>>;
+  rmm::device_uvector<stack_value_type> stack_storage(orc_columns.size(), stream);
 
+  // pre-order append ORC device columns
   cudf::detail::device_single_thread(
-    [d_orc_cols = device_span<orc_column_device_view>{d_orc_columns},
-     d_table    = d_table] __device__() mutable {
+    [d_orc_cols         = device_span<orc_column_device_view>{d_orc_columns},
+     d_table            = d_table,
+     stack_storage      = stack_storage.data(),
+     stack_storage_size = stack_storage.size()] __device__() {
+      device_stack stack(stack_storage, stack_storage_size);
```

Comment:
I don't know how much performance is of concern here, but using shared memory for the backing store of the stack would certainly be faster.